eservice/lib/libpdo_enclave/contract_enclave.cpp

```diff
@@ -41,6 +42,119 @@
 
 ByteArray last_result;
+ContractWorker *worker = NULL;
+static bool worker_initialized = false;
+static bool shutdown_worker = false;
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+pdo_err_t ecall_CreateContractWorker(size_t inThreadId){
+    pdo_err_t result = PDO_SUCCESS;
+
+    try {
+        if (!worker_initialized)
+        {
+            printf("ThreadID: %ld - Creating new ContractWorker\n", (long) inThreadId);
+            worker = new ContractWorker((long) inThreadId);
+
+            worker_initialized = true;
+        }
+
+        while (!shutdown_worker)
+        {
+            worker->interpreterInit();
+        }
+    }
+    catch (pdo::error::Error& e)
+    {
+        printf("Error in contract enclave (ecall_CreateContractWorker): %04X -- %s \n", e.error_code(), e.what());
+        ocall_SetErrorMessage(e.what());
+        result = e.error_code();
+    }
+    catch (...)
+    {
+        printf("Unknown error in contract enclave (ecall_CreateContractWorker)\n");
+        result = PDO_ERR_UNKNOWN;
+    }
+
+    return result;
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+pdo_err_t ecall_ShutdownContractWorker(){
+    pdo_err_t result = PDO_SUCCESS;
+
+    try {
+        if (worker_initialized)
+        {
+            printf("ThreadID: %ld - Shutting down ContractWorker\n", (long) worker->threadId_);
+            shutdown_worker = true;
+        }
+    }
+    catch (pdo::error::Error& e)
+    {
+        printf("Error in contract enclave (ecall_ShutdownContractWorker): %04X -- %s \n", e.error_code(), e.what());
+        ocall_SetErrorMessage(e.what());
+        result = e.error_code();
+        ocall_SetErrorMessage(e.what());
+        result = e.error_code();
+    }
+    catch (...)
+    {
+        printf("Unknown error in contract enclave (ecall_ShutdownContractWorker)");
+        result = PDO_ERR_UNKNOWN;
+    }
+
+    return result;
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+ContractWorker::ContractWorker(long threadId)
+{
+    this->threadId_ = threadId;
+    this->current_state_ = INTERPRETER_DONE;
+    sgx_thread_mutex_init(&this->mutex_, NULL);
+    this->interpreterInit();
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+void ContractWorker::interpreterInit()
+{
+    sgx_thread_mutex_lock(&this->mutex_);
+
+    if (current_state_ == INTERPRETER_DONE) {
+        if (interpreter_ != NULL) {
+            delete interpreter_;
+        }
+
+        interpreter_ = new GipsyInterpreter();
+
+        current_state_ = INTERPRETER_READY;
+        sgx_thread_cond_signal(&this->ready_cond_);
+    }
+
+    sgx_thread_mutex_unlock(&this->mutex_);
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+void ContractWorker::interpreterIsReady()
+{
+
+    while (!this->current_state_ == INTERPRETER_READY){
+        sgx_thread_cond_wait(&this->ready_cond_, &this->mutex_);
```

Comment:
This needs to lock the mutex before checking the current state. If it isn't ready, then it needs to unlock, then wait on the signal.